### PR TITLE
Implement non-chunked parser

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

File: core/src/Streamly/Data/ParserK.hs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+-- Copyright   : (c) 2023 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : pre-release
+-- Portability : GHC
+--
+-- Parsers using Continuation Passing Style (CPS). See notes in
+-- "Streamly.Data.Parser" module to know when to use this module.
+--
+-- To run a 'ParserK' use 'Streamly.Data.StreamK.parseBreak'.
```
Comment (adithyaov) [inline core/src/Streamly/Data/ParserK.hs:12]:
```suggestion
-- To run a 'ParserK' use 'Streamly.Data.StreamK.parse'.
```

File: core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+    , ParserK (..)
+    , fromParser
+    -- , toParser
+    , fromPure
+    , fromEffect
+    , die
+    )
+where
+
+#include "ArrayMacros.h"
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs:38]:
```suggestion
```

File: core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+import Control.Applicative (Alternative(..), liftA2)
+import Control.Monad (MonadPlus(..), ap)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+-- import Control.Monad.Trans.Class (MonadTrans(lift))
+import GHC.Types (SPEC(..))
+
+import qualified Control.Monad.Fail as Fail
+import qualified Streamly.Internal.Data.Parser.ParserD.Type as ParserD
+
+data Input a = None | Chunk a
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs:51]: Rename this to `Single`

File: core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+-- an error.
+--
+-- See ParserD docs. This is the same as the ParserD Step except that it uses a
+-- continuation in Partial and Continue constructors instead of a state in case
+-- of ParserD.
+--
+-- /Pre-release/
+--
+data Step a m r =
+    -- The Int is the current stream position index wrt to the start of the
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs:64]: Change docs

File: core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+    fmap f (Continue n k) = Continue n (fmap (fmap f) . k)
+    fmap _ (Error e) = Error e
+
+-- Note: Passing position index separately instead of passing it with the
+-- result causes huge regression in expression parsing becnhmarks.
+
+-- | The parser's result.
+--
+-- Int is the position index into the current input array. Could be negative.
+-- Cannot be beyond the input array max bound.
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs:86]: Change docs

File: core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+--
+-- /Pre-release/
+--
+data ParseResult b =
+      Success !Int !b     -- Leftover count, result
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs:91]: Not leftover.

File: core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+    | Failure !String     -- Error
+
+-- | Map a function over 'Success'.
+instance Functor ParseResult where
+    fmap f (Success n b) = Success n (f b)
+    fmap _ (Failure e) = Failure e
+
+-- XXX Change the type to the shape (a -> m r -> m r) -> (m r -> m r) -> m r
+--
+-- The parse continuation would be: Array a -> m (Step a m r) -> m (Step a m r)
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs:101]:
```suggestion
-- The parse continuation would be: a -> m (Step a m r) -> m (Step a m r)
```

File: core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+           -- for cases where we process just one element per call.
+           --
+           -- Do not eta reduce the applications of this continuation.
+           --
+           (ParseResult b -> Int -> Input a -> m (Step a m r))
+           -- XXX Maintain and pass the original position in the stream. that
+           -- way we can also report better errors. Use a Context structure for
+           -- passing the state.
+
+           -- Stream position index wrt to the current input array start. If
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs:126]: Docs,

File: core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs [hunk 9]
```diff
@@ -0,0 +1,500 @@
...
+                ParserD.Error err ->
+                    cont (Failure err) count (Chunk x)
+
+    {-# NOINLINE parseContNothing #-}
+    parseContNothing !count !pst = do
+        r <- extract pst
+        case r of
+            -- IMPORTANT: the n here is from the byte stream parser, that means
+            -- it is the backtrack element count and not the stream position
+            -- index into the current input array.
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs:425]: Change

File: core/docs/Changelog.md [hunk 10]
```diff
@@ -1,5 +1,12 @@
 # Changelog
 
+## Unreleased
+
+* `Streamly.Data.ParserK` now has a non-chunked parser.
+  * The chunked parser is moved to `Streamly.Data.ParserK.Chunked`
+  * `ParserK` is now representes the non-chunked parser. `ChunkParserK`
```
Comment (harendra-kumar) [inline core/docs/Changelog.md:7]: "now represents"

File: core/docs/Changelog.md [hunk 11]
```diff
@@ -1,5 +1,12 @@
 # Changelog
 
+## Unreleased
+
+* `Streamly.Data.ParserK` now has a non-chunked parser.
```
Comment (harendra-kumar) [inline core/docs/Changelog.md:5]: "is now"

File: core/src/Streamly/Internal/Data/Stream/StreamK.hs [hunk 12]
```diff
@@ -1380,6 +1406,115 @@ parseChunks :: (Monad m, Unbox a) =>
     ChunkParserK a m b -> StreamK m (Array a) -> m (Either ParseError b)
 parseChunks f = fmap fst . parseBreakChunks f
 
+-- | Run a 'ParserK' over a chunked 'StreamK' and return the rest of the Stream.
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Stream/StreamK.hs:1409]: Please use `parseBreakChunks` where possible, for better performance.

File: core/src/Streamly/Internal/Data/Stream/StreamK.hs [hunk 13]
```diff
@@ -1380,6 +1406,115 @@ parseChunks :: (Monad m, Unbox a) =>
...
+        -> StreamK m a
+        -> m (Either ParseError b, StreamK m a)
+    go backBuf parserk stream = do
+        let stop = goStop backBuf parserk
+            single a = yieldk backBuf parserk a nil
+         in foldStream
+                defState (yieldk backBuf parserk) single stop stream
+
+{-# INLINE parse #-}
+parse :: Monad m =>
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Stream/StreamK.hs:1514]: Document this.

Please use `parseChunks` where possible, for better performance.

**Comment by adithyaov** [comment]
See https://github.com/composewell/streamly/pull/2502

**Pull request closed without merging**